### PR TITLE
add hangman game

### PR DESCRIPTION
**Pull request by namemcguffin**
Added hangman game

## Description
added hangman game that can be invoked with either `?hm` or `?hangman` for a random category word
or with `?hm x` or `?hangman x` where` x` is one of the valid categories
also has a help message, callable with `?hm help` or `?hangman help`

additions:

+ `hangman.py` in `./cogs/utils/`, which is used to generate the hangman word lists, and which contains the `HangmanState` class used in the hangman command
+ `hangman_dict.obj` in `./data/premade/` directory, which contains word lists for the hangman game
+ `hangman` command to `games.py`
+ `HANGMAN_REWARD` action to `currency.py` 

changes:

+ renamed `requests.py` in `./cogs/utils/` to `custom_requests.py` so that it doesn't conflict when importing `requests` (which is needed in `hangman.py`)
+ changed imports in `helpers.py` (due to above change)

## Motivation and Context
ported missing functionality from Mose II 

## How Has This Been Tested?
tested extensively in the idoneam discord

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: cogs/utils/hangman.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        word_dict[link_name] = (word_list,
+                                sub(r"[^a-z\",'/ ]", "",
+                                    link.contents[0].lower()))
+        print(word_dict[link_name])
+    with open(file_name, "wb") as dump_file:
+        dump(word_dict, dump_file)
+
+
+if __name__ == "__main__":
+    mk_word_dict("hangman_dict")
```
Comment (davidlougheed) [inline cogs/utils/hangman.py:81]: i'd prefer if this had the `.pickle` extension

File: cogs/utils/hangman.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            get("https://www.enchantedlearning.com/wordlist/").content,
+            "html.parser").find_all("table")[-6].find_all(
+                "a", {"target": "_top"})
+        if search(r"^/wordlist/([a-z]*)\.shtml$", elem["href"])
+    ]
+    word_dict = {}
+    for link in link_list:
+        print(link["href"], sub(r"[^a-z\",'/ ]", "", link.contents[0].lower()))
+        link_name = findall(r"^/wordlist/([a-z]*)\.shtml$", link["href"])[0]
+        word_list = []
```
Comment (davidlougheed) [inline cogs/utils/hangman.py:66]:
```python
word_list = [
    word.text.lower() 
    for word in BeautifulSoup(
                get(f"https://www.enchantedlearning.com/{link['href']}").
                content, "html.parser").find_all("div",
                                                 {"class": "wordlist-item"})
]
```

File: cogs/games.py [hunk 3]
```diff
@@ -20,18 +20,203 @@
...
+                                               check=same_channel_check,
+                                               timeout=300)
+            curr_guess = curr_msg.content
+            if not (curr_msg.author in timeout_dict and
+                    (time() - timeout_dict[curr_msg.author]) < 3.0):
+                if curr_guess in "abcdefghijklmnopqrstuvwxyz" and len(
+                        curr_guess) == 1:
+                    await curr_msg.delete()
+                    if curr_guess in not_guessed:    # curr_guess in not_guessed => curr_guess is correct and new
+                        not_guessed.remove(curr_guess)
```
Comment (davidlougheed) [inline cogs/games.py:102]: this gets a bit uncomfortably indented, i wonder if there's a way with moving around the if/else cases and using `continue` or something to make it less nested

File: cogs/games.py [hunk 4]
```diff
@@ -20,18 +20,203 @@
...
+        except KeyError:
+            await ctx.send(
+                f"invalid category command, here is a list of valid commands: {sorted(self.hangman_dict.keys())}"
+            )
+            return
+        num_mistakes = 0
+        not_guessed = set(re.sub(r"[^a-z]", "", word))
+        incorrect_guesses = set()
+        first_line = "".join(char + " " if char not in not_guessed else "_ "
+                             for char in word).rstrip()
```
Comment (davidlougheed) [inline cogs/games.py:77]: instead of adding spaces here + rstrip, you could just join by space
```python
first_line = " ".join(char if char not in not_guessed else "_" for char in word)
```

File: cogs/games.py [hunk 5]
```diff
@@ -20,18 +20,203 @@
...
+            curr_guess = curr_msg.content
+            if not (curr_msg.author in timeout_dict and
+                    (time() - timeout_dict[curr_msg.author]) < 3.0):
+                if curr_guess in "abcdefghijklmnopqrstuvwxyz" and len(
+                        curr_guess) == 1:
+                    await curr_msg.delete()
+                    if curr_guess in not_guessed:    # curr_guess in not_guessed => curr_guess is correct and new
+                        not_guessed.remove(curr_guess)
+                        first_line = "".join(
+                            char + " " if char not in not_guessed else "_ "
```
Comment (davidlougheed) [inline cogs/games.py:104]: space join thing again; maybe could be deduplicated into a helper function?

File: cogs/games.py [hunk 6]
```diff
@@ -20,18 +20,203 @@
...
+                        if len(player_msg_list) > 3:
+                            player_msg_list = player_msg_list[-3:]
+                        txt_embed.set_field_at(
+                            0,
+                            name=f"hangman (category: {cat_name})",
+                            value=
+                            f"`{first_line}`\n```{HANG_LIST[num_mistakes]}```\n{NEWLINE.join(player_msg_list)}"
+                        )
+                        await hg_msg.edit(embed=txt_embed)
+                        if len(not_guessed) == 0:
```
Comment (davidlougheed) [inline cogs/games.py:118]: this can be
```python
if not not_guessed:
```

File: cogs/games.py [hunk 7]
```diff
@@ -20,18 +20,203 @@
...
+                        )
+                        await hg_msg.edit(embed=txt_embed)
+                    elif curr_guess not in incorrect_guesses:    # curr_guess not in not_guessed and not in word (elif) and not in incorrect guesses => curr_guess is incorrect and new
+                        num_mistakes += 1
+                        incorrect_guesses.add(curr_guess)
+                        last_line = f"incorrect guesses: {str(sorted(incorrect_guesses))[1:-1]}"
+                        player_msg_list.append(
+                            f"{curr_msg.author} guessed '{curr_guess}' wrong!")
+                        if len(player_msg_list) > 3:
+                            player_msg_list = player_msg_list[-3:]
```
Comment (davidlougheed) [inline cogs/games.py:142]: this code seems to be repeated a lot, so if it could be factored out maybe that would be useful?

File: cogs/games.py [hunk 8]
```diff
@@ -20,18 +20,203 @@
...
+                        timeout_dict[curr_msg.author] = time()
+                        txt_embed.set_field_at(
+                            0,
+                            name=f"hangman (category: {cat_name})",
+                            value=
+                            f"`{first_line}`\n```{HANG_LIST[num_mistakes]}```\n{NEWLINE.join(player_msg_list)}"
+                        )
+                        txt_embed.set_footer(text=last_line)
+                        await hg_msg.edit(embed=txt_embed)
+                        if num_mistakes == 6:
```
Comment (davidlougheed) [inline cogs/games.py:152]: maybe make this 6 a global variable since it feels like a magic number rn

File: cogs/games.py [hunk 9]
```diff
@@ -20,18 +20,203 @@
...
+                            await ctx.send(
+                                f"sorry everyone, {curr_msg.author} used your last chance, the right answer was `{word}`"
+                            )
+                    else:    # curr_guess not in not_guessed and not in word and in incorrect_guesses (else) => curr_guess is incorrect but already made
+                        timeout_dict[curr_msg.author] = time()
+                        player_msg_list.append(
+                            f"{curr_msg.author}, '{curr_guess}' was already guessed"
+                        )
+                        if len(player_msg_list) > 3:
+                            player_msg_list = player_msg_list[-3:]
```
Comment (davidlougheed) [inline cogs/games.py:162]: `> 3` thing again
also i just realized this if statement isnt needed at all, since e.g. `["a", "b"][-3:]` is just `["a", "b"]`

File: cogs/games.py [hunk 10]
```diff
@@ -20,18 +20,203 @@
...
+                        txt_embed.set_field_at(
+                            0,
+                            name=f"hangman (category: {cat_name})",
+                            value=
+                            f"`{first_line}`\n```{HANG_LIST[num_mistakes]}```\n{NEWLINE.join(player_msg_list)}"
+                        )
+                        await hg_msg.edit(embed=txt_embed)
+                elif curr_guess == word:
+                    not_guessed = set()
+                    first_line = "".join(char + " " for char in word).rstrip()
```
Comment (davidlougheed) [inline cogs/games.py:172]: space join thing again

File: cogs/games.py [hunk 11]
```diff
@@ -20,18 +20,203 @@
...
+                    player_msg_list.append(
+                        f"{curr_msg.author} guessed the entire word ('{word}') correctly!"
+                    )
+                    if len(player_msg_list) > 3:
+                        player_msg_list = player_msg_list[-3:]
```
Comment (davidlougheed) [inline cogs/games.py:177]: list truncate thing again

File: cogs/games.py [hunk 12]
```diff
@@ -20,18 +20,203 @@
...
+                    await ctx.send(
+                        f"congratulations {curr_msg.author}, you solved the hangman, but in a cool way"
+                    )
+                elif len(curr_guess) != 0:
+                    invalid_msg_count += 1
+                    player_msg_list.append(
+                        f"{curr_msg.author}, invalid guess"
+                    )
+                    if len(player_msg_list) > 3:
+                        player_msg_list = player_msg_list[-3:]
```
Comment (davidlougheed) [inline cogs/games.py:194]: etc

File: cogs/games.py [hunk 13]
```diff
@@ -20,18 +20,203 @@
...
+                        invalid_msg_count = 0
+                        await hg_msg.delete()
+                        hg_msg = await ctx.send(embed=hg_msg.embeds[0])
+            elif curr_guess in "abcdefghijklmnopqrstuvwxyz" and len(
+                    curr_guess) == 1:
+                await curr_msg.delete()
+                player_msg_list.append(
+                    f"{curr_msg.author} you cannot guess right now!")
+                if len(player_msg_list) > 3:
+                    player_msg_list = player_msg_list[-3:]
```
Comment (davidlougheed) [inline cogs/games.py:212]: etc

**Review by davidlougheed — CHANGES_REQUESTED** [review]

File: cogs/games.py [hunk 14]
```diff
@@ -20,18 +20,203 @@
...
+        """
+        if command == "help":
+            await ctx.send(
+                f"rules: 6 wrong guesses are allowed, guesses must be either the entire correct word or a single lowercase letter\nhere is a list of valid category commands: {sorted(self.hangman_dict.keys())}"
+            )
+            return
+        if command is None:
+            category = choice(list(self.hangman_dict.keys()))
+        else:
+            category = command
```
[the remark this answers is not in the document]
Comment (namemcguffin) [reply cogs/games.py]: good golf, changed

File: cogs/utils/hangman.py (hunk 2 above)
Comment (namemcguffin) [reply cogs/utils/hangman.py]: ah yes, python list comprehension, god its so good
(changed)

File: cogs/utils/hangman.py (hunk 1 above)
Comment (namemcguffin) [reply cogs/utils/hangman.py:79]: yep, added

File: cogs/games.py (hunk 3 above)
Comment (namemcguffin) [reply cogs/games.py:99]: i honestly dont know how, but ill try to think about it, see if i can come up with something

File: cogs/games.py (hunk 4 above)
Comment (namemcguffin) [reply cogs/games.py]: that rstrip call bothered me so much, thats a really good golf, i knew i was missing something
thx! (changed)

File: cogs/games.py (hunk 9 above)
Comment (namemcguffin) [reply cogs/games.py:157]: good point, added

File: cogs/games.py (hunk 8 above)
Comment (namemcguffin) [reply cogs/games.py]: good catch, fixed

File: cogs/games.py [hunk 15]
```diff
@@ -20,18 +20,203 @@
...
+                        player_msg_list = player_msg_list[-3:]
+                    txt_embed.set_field_at(
+                        0,
+                        name=f"hangman (category: {cat_name})",
+                        value=
+                        f"`{first_line}`\n```{HANG_LIST[num_mistakes]}```\n{NEWLINE.join(player_msg_list)}"
+                    )
+                    await hg_msg.edit(embed=txt_embed)
+                    await ctx.send(
+                        f"congratulations {curr_msg.author}, you solved the hangman, but in a cool way"
```
Comment (le-potate) [inline cogs/games.py:179]: i wouldn't write the "but in a cool way" part, this is how most hangmans are solved. also, write "the answer was [...]"

File: cogs/games.py [hunk 16]
```diff
@@ -20,18 +20,203 @@
...
+                    )
+                elif len(curr_guess) != 0:
+                    invalid_msg_count += 1
+                    player_msg_list.append(
+                        f"{curr_msg.author}, invalid guess"
```
Comment (le-potate) [inline cogs/games.py]: Don't write Invalid guess everytime, this is gonna be annoying when unrelated messages are sent (or sometimes people just, talk about the hangman too, and theyll get hit with a ton of "invalid guess")

File: cogs/games.py [hunk 17]
```diff
@@ -20,18 +20,203 @@
...
+                        0,
+                        name=f"hangman (category: {cat_name})",
+                        value=
+                        f"`{first_line}`\n```{HANG_LIST[num_mistakes]}```\n{NEWLINE.join(player_msg_list)}"
+                    )
+                    await hg_msg.edit(embed=txt_embed)
+                    await ctx.send(
+                        f"congratulations {curr_msg.author}, you solved the hangman, but in a cool way"
+                    )
+                elif len(curr_guess) != 0:
```
Comment (le-potate) [inline cogs/games.py:181]: im not sure a message of 0 length can be sent, but anyways this case is useless if you change the check like ive written elsewhere

File: cogs/games.py [hunk 18]
```diff
@@ -20,18 +20,203 @@
...
+        not_guessed = set(re.sub(r"[^a-z]", "", word))
+        incorrect_guesses = set()
+        first_line = "".join(char + " " if char not in not_guessed else "_ "
+                             for char in word).rstrip()
+        last_line = "incorrect guesses: "
+        player_msg_list = []
+        timeout_dict = {}
+        invalid_msg_count = 0
+        same_channel_check = lambda msg: msg.channel == ctx.message.channel
+        txt_embed = Embed(colour=0xFF0000)
```
Comment (le-potate) [inline cogs/games.py:81]: use discord.Embed

File: cogs/games.py [hunk 19]
```diff
@@ -20,18 +20,203 @@
 # discord-py requirements
 import discord
 from discord.ext import commands
+from discord import Embed
```
Comment (le-potate) [inline cogs/games.py:23]: I wouldn't do this, you only use Embed once in your code so its fine to use discord.Embed for it, and this is how it is done in every other cog. embed is also a commonly used variable name so this would be a bit confusing when people write other commands in this cog

File: cogs/games.py [hunk 20]
```diff
@@ -20,18 +20,203 @@
...
+                            0,
+                            name=f"hangman (category: {cat_name})",
+                            value=
+                            f"`{first_line}`\n```{HANG_LIST[num_mistakes]}```\n{NEWLINE.join(player_msg_list)}"
+                        )
+                        txt_embed.set_footer(text=last_line)
+                        await hg_msg.edit(embed=txt_embed)
+                        if num_mistakes == 6:
+                            await ctx.send(
+                                f"sorry everyone, {curr_msg.author} used your last chance, the right answer was `{word}`"
```
Comment (le-potate) [inline cogs/games.py:149]: also update the embed

File: cogs/games.py [hunk 21]
```diff
@@ -20,18 +20,203 @@
...
+                        txt_embed.set_field_at(
+                            0,
+                            name=f"hangman (category: {cat_name})",
+                            value=
+                            f"`{first_line}`\n```{HANG_LIST[num_mistakes]}```\n{NEWLINE.join(player_msg_list)}"
+                        )
+                        await hg_msg.edit(embed=txt_embed)
+                        if len(not_guessed) == 0:
+                            await ctx.send(
+                                f"congratulations {curr_msg.author}, you solved the hangman"
```
Comment (le-potate) [inline cogs/games.py:117]: update the embed

**Review by le-potate — CHANGES_REQUESTED** [review]

File: cogs/games.py [hunk 22]
```diff
@@ -20,18 +20,203 @@
...
+            value=
+            f"`{first_line}`\n```{HANG_LIST[num_mistakes]}```\n{NEWLINE.join(player_msg_list)}"
+        )
+        txt_embed.set_footer(text=last_line)
+        hg_msg = await ctx.send(embed=txt_embed)
+        while len(not_guessed) > 0 and num_mistakes < 6:
+            curr_msg = await self.bot.wait_for('message',
+                                               check=same_channel_check,
+                                               timeout=300)
+            curr_guess = curr_msg.content
```
[the remark this answers is not in the document]
Comment (namemcguffin) [reply cogs/games.py]: good point, fixed it

File: cogs/games.py [hunk 23]
```diff
@@ -20,18 +20,203 @@
...
+        txt_embed.add_field(
+            name=f"hangman (category: {cat_name})",
+            value=
+            f"`{first_line}`\n```{HANG_LIST[num_mistakes]}```\n{NEWLINE.join(player_msg_list)}"
+        )
+        txt_embed.set_footer(text=last_line)
+        hg_msg = await ctx.send(embed=txt_embed)
+        while len(not_guessed) > 0 and num_mistakes < 6:
+            curr_msg = await self.bot.wait_for('message',
+                                               check=same_channel_check,
```
[the remark this answers is not in the document]
Comment (namemcguffin) [reply cogs/games.py]: thats a damn good idea yea (implemented in new commit)

File: cogs/games.py (hunk 15 above)
Comment (namemcguffin) [reply cogs/games.py]: changed it to a rarer message in new commit

File: cogs/games.py [hunk 24]
```diff
@@ -20,18 +20,203 @@
...
 class Games(commands.Cog):
-    def __init__(self, bot):
+    def __init__(self, bot, hangman_tbl_name: str):
         self.bot = bot
+        my_path = os.path.abspath(os.path.dirname(__file__))
+        hangman_path = os.path.join(my_path, f"utils/{hangman_tbl_name}")
+        with open(hangman_path, "rb") as hangman_pkl:
+            self.hangman_dict: Dict[str, Tuple[str, str]] = load(hangman_pkl)
+
+    @commands.max_concurrency(1, per=commands.BucketType.channel, wait=False)
```
[the remark this answers is not in the document]
Comment (le-potate) [reply cogs/games.py:49]: Done in  #359

**Comment by davidlougheed** [comment]
we should integrate currency as a reward for this!

**Review by davidlougheed — CHANGES_REQUESTED** [review]
this code gets very nested and theres some repetitive stuff which should be reduced

File: cogs/currency.py [hunk 25]
```diff
@@ -112,6 +114,15 @@ async def create_bank_transaction(self, c, user: discord.Member,
             "INSERT INTO BankTransactions(UserID, Amount, Action, "
             "Metadata, Date) VALUES(?, ?, ?, ?, ?)", t)
 
+    async def give_user_cheeps(self, ctx, amount: Decimal, action: str,
```
Comment (davidlougheed) [inline cogs/currency.py:117]: the bot doesnt have an explicit name for currency
also this function isnt needed, use `create_bank_transaction`

File: cogs/games.py [hunk 26]
```diff
@@ -23,15 +23,230 @@
...
+from asyncio import TimeoutError
+from typing import Dict, List, Set, Tuple
 from .utils.dice_roll import dice_roll
 from .utils.clamp_default import clamp_default
+from .utils.hangman import LOSS_MISTAKES, mk_hangman_str
+from .currency import HANGMAN_REWARD
 
 ROLL_PATTERN = re.compile(r'^(\d*)d(\d*)([+-]?\d*)$')
+HM_WIN_CHEEPS = 10
+HM_COOL_WIN_CHEEPS = 20
```
Comment (davidlougheed) [inline cogs/games.py:39]: these 2 should ideally be put in config.ini along with other currency-tweaking params

File: cogs/games.py [hunk 27]
```diff
@@ -23,15 +23,230 @@
...
+                            0,
+                            name=f"hangman (category: {pretty_name})",
+                            value=mk_hangman_str(first_line, num_mistakes,
+                                                 "\n".join(player_msg_list)))
+                        await ctx.send(embed=hm_embed)
+                        await ctx.send(
+                            f"sorry everyone, {curr_msg.author} used your last chance, the right answer was `{hm_word}`"
+                        )
+                        break
+                else:    # curr_guess not in not_guessed and not in word and in incorrect_guesses (else) => curr_guess is incorrect but already made
```
Comment (davidlougheed) [inline cogs/games.py:220]: move comment into else block and wrap it to line limits

File: cogs/games.py [hunk 28]
```diff
@@ -23,15 +23,230 @@
...
+                        append_and_slice(
+                            f"{curr_msg.author} used your last chance!")
+                        hm_embed.set_field_at(
+                            0,
+                            name=f"hangman (category: {pretty_name})",
+                            value=mk_hangman_str(first_line, num_mistakes,
+                                                 "\n".join(player_msg_list)))
+                        await ctx.send(embed=hm_embed)
+                        await ctx.send(
+                            f"sorry everyone, {curr_msg.author} used your last chance, the right answer was `{hm_word}`"
```
Comment (davidlougheed) [inline cogs/games.py:217]: wrap to line limits

**Review by davidlougheed — CHANGES_REQUESTED** [review]

File: cogs/games.py [hunk 29]
```diff
@@ -23,15 +23,182 @@
...
+                    winner = curr_msg.author
+                    cool_win = len(not_guessed) > (len(set(lowered_word)) //
+                                                   2.5)
+                    update_embed(f"{winner} guessed the entire word!",
+                                 correct_guess=True,
+                                 img_url=hm_img)
+                    await ctx.send(embed=hm_embed)
+                    await ctx.send(
+                        f"congratulations {winner}, you solved the hangman"
+                        f"{', but in a cool way' if  cool_win else ''}, "
```
Comment (davidlougheed) [inline cogs/games.py:134]: "but in a cool way" i'd say isn't ideal grammar here - maybe "[...] you solved the hangman (in a cool way 😎), earning you [...]" or something like that

File: cogs/subscribers.py [hunk 30]
```diff
@@ -34,7 +34,6 @@
 # Subscriber decorator
 from .utils.subscribers import canary_subscriber
 
-CFIA_FEED_URL = "http://inspection.gc.ca/eng/1388422350443/1388422374046.xml"
```
Comment (davidlougheed) [inline cogs/subscribers.py:37]: why is this removed?

**Review by davidlougheed — CHANGES_REQUESTED** [review]

File: cogs/subscribers.py (hunk 30 above)
Comment (namemcguffin) [reply cogs/subscribers.py:37]: i think it was a typo or something, i dont think it was intended

File: cogs/utils/hangman.py [hunk 31]
```diff
@@ -0,0 +1,239 @@
+# -*- coding: utf-8 -*-
```
Comment (davidlougheed) [inline cogs/utils/hangman.py:1]: Remove the first two lines

**Review by davidlougheed — CHANGES_REQUESTED** [review]

**Review by davidlougheed — CHANGES_REQUESTED** [review]

File: cogs/utils/hangman.py [hunk 32]
```diff
@@ -0,0 +1,241 @@
...
+        self.num_mistakes: int = 0
+        self.embed = discord.Embed(
+            colour=random.randint(0, 0xFFFFFF)
+        ).add_field(
+            name=self.field_name,
+            value=f"`{self.first_line}`\n```{HANG_LIST[self.num_mistakes]}```"
+        ).set_footer(text=self.last_line)
+
+    def full(self):
+        self.first_line = " ".join(char for char in self.word)
```
Comment (davidlougheed) [inline cogs/utils/hangman.py:196]: word already iterable - replace with
```python
" ".join(self.word)
```

File: cogs/games.py [hunk 33]
```diff
@@ -23,15 +23,174 @@
...
+                game_state.add_msg(
+                    f"{curr_msg.author} you cannot guess right now!")
+                await ctx.send(embed=game_state.embed)
+        if winner is not None:
+            conn = sqlite3.connect(self.bot.config.db_path)
+            await self.bot.get_cog("Currency").create_bank_transaction(
+                conn.cursor(), winner,
+                self.hm_cool_win if cool_win else self.hm_norm_win,
+                HANGMAN_REWARD, {"cool": cool_win})
+            conn.commit()
```
Comment (davidlougheed) [inline cogs/games.py:193]: missing `conn.close()` (still would implicitly close, but better to be explicit here)

File: cogs/games.py [hunk 34]
```diff
@@ -23,15 +23,174 @@
...
+                            f"{curr_msg.author} used your last chance!")
+                        if game_state.img:
+                            game_state.embed.set_image(url=game_state.img)
+                        await ctx.send(embed=game_state.embed)
+                        await ctx.send(
+                            f"sorry everyone, `{curr_msg.author}` used your "
+                            f"last chance, the right answer was `{game_state.word}`"
+                        )
+                        break
+            else:
```
Comment (davidlougheed) [inline cogs/games.py:183]: flip condition and early-continue with this else to reduce indentation / put "non-normal" conditions earlier in loop logic

File: cogs/games.py [hunk 35]
```diff
@@ -23,15 +23,174 @@
...
+        if command in ["movies", "kino"]:
+            category = "movie"
+        elif command in ["elements"]:
+            category = "element"
+        elif command in ["countries"]:
+            category = "country"
+        elif command in ["animals"]:
+            category = "animal"
+        else:
+            category: str = command or random.choice(
```
Comment (davidlougheed) [inline cogs/games.py:82]:
```python
CATEGORY_SYNONYMS = {
    "movies": "movie",
    "kino": "movie",
    "elements": "element",
    "countries": "country",
    "animals": "animal",
}

# ...

# Translate category name synonyms; if no category specified, choose one at random
category = CATEGORY_SYNONYMS.get(command, command) or \
    random.choice(list(self.hangman_dict.keys())))
```

File: cogs/games.py [hunk 36]
```diff
@@ -23,15 +23,174 @@
...
+            cat_list: str = ", ".join(
+                f"`{hm_cat}` (length: {len(self.hangman_dict[hm_cat][0])})"
+                for hm_cat in sorted(self.hangman_dict.keys()))
+            await ctx.send(
+                f"rules: {LOSS_MISTAKES - 1} wrong guesses are allowed, "
+                f"guesses must be either the entire correct word or a "
+                f"single letter (interpreted in a case insensitive manner)\n"
+                f"here is a list of valid category commands: {cat_list}")
+            return
+        if command in ["movies", "kino"]:
```
Comment (davidlougheed) [inline cogs/games.py:73]: add whitespace between this early-return if and the other conditionals/logic

File: cogs/games.py [hunk 37]
```diff
@@ -23,15 +23,174 @@
...
+        self.hm_norm_win = bot.config.games["hm_norm_win"]
+        self.hm_timeout = bot.config.games["hm_timeout"]
+        with open(f"{os.getcwd()}/data/premade/{hangman_tbl_name}.obj",
+                  "rb") as hangman_pkl:
+            self.hangman_dict: dict[str, tuple[list[tuple[str, Optional[str]]],
+                                               str]] = pickle.load(hangman_pkl)
+
+    @commands.max_concurrency(1, per=commands.BucketType.channel, wait=False)
+    @commands.command(aliases=["hm"])
+    async def hangman(self, ctx, *, command: str = None):
```
Comment (davidlougheed) [inline cogs/games.py:54]: `command: Optional[str]`

File: cogs/games.py [hunk 38]
```diff
@@ -23,15 +23,174 @@
...
+                    if game_state.img:
+                        game_state.embed.set_image(url=game_state.img)
+                    await ctx.send(embed=game_state.embed)
+                    await ctx.send(
+                        f"congratulations `{winner}`, you solved the hangman" +
+                        (f" (in a cool way), earning you {self.hm_cool_win} cheeps"
+                         if cool_win else
+                         f", earning you {self.hm_norm_win} cheeps"))
+                    break
+                if curr_guess in game_state.previous_guesses:
```
Comment (davidlougheed) [inline cogs/games.py:140]: more whitespace between blocks for all these big conditionals to improve readability

**Review by davidlougheed — DISMISSED** [review]

**Review by le-potate — CHANGES_REQUESTED** [review]
I've already tested and reviewed this directly on the server while it was being written, so I don't think there's any major bugs remaining. After this I think it's ready to merge now

File: cogs/games.py [hunk 39]
```diff
@@ -23,15 +23,177 @@
...
+                game_state.add_msg("the game has timed out")
+                await ctx.send(embed=game_state.embed)
+                await ctx.send(
+                    f"sorry everyone, no one has interacted with the "
+                    f"hangman in {self.hm_timeout//60} minutes, "
+                    f"the game has timed out")
+                return
+
+            curr_guess = curr_msg.content.lower()
+
```
[the remark this answers is not in the document]
Comment (namemcguffin) [reply cogs/games.py:122]: i dont want to use just "quit", so that a conversation doesn't proc it, but that's a good point, so ill use `?hangman quit` instead?

[the remark this answers is not in the document]
Comment (le-potate) [reply cogs/games.py:122]: I think people sending a message that just says "quit" is pretty unlikely, not like that couldnt happen with the answers too.
But if you want to just do `?hangman quit` it's fine too. Just like I said in the other comment for `help` you'd need to deal with max concurrency too

File: cogs/games.py [hunk 40]
```diff
@@ -23,15 +23,225 @@
...
+                                               str]] = pickle.load(hangman_pkl)
+
+    def help_str(self):
+        cat_list: str = ", ".join(
+            f"`{hm_cat}` (length: {len(self.hangman_dict[hm_cat][0])})"
+            for hm_cat in sorted(self.hangman_dict.keys()))
+        return ("rules: 5 wrong guesses are allowed, guesses "
+                "must be either the entire correct word or a "
+                "single letter (interpreted in a case insensitive manner)\n"
+                f"here is a list of valid category commands: {cat_list}\n"
```
Comment (le-potate) [inline cogs/games.py:68]: I would call this a category argument

File: cogs/games.py [hunk 41]
```diff
@@ -23,15 +23,225 @@
...
+        return msg.channel == hm_channel and (
+            (len(msg.content) == 1 and msg.content.isalpha())
+            or msg.content.lower() == lowered)
+
+    @commands.command(aliases=["hm"])
+    async def hangman(self, ctx, command: Optional[str] = None):
+        """
+        Play a nice game of hangman with internet strangers!
+        Guesses must be single letters (interpreted in a case insensitive manner) or the entire correct word.
+        Can either be called with "?{hm|hangman}" or "?{hm|hangman} x", where x is a valid category command.
```
Comment (le-potate) [inline cogs/games.py:83]: category argument

File: cogs/games.py [hunk 42]
```diff
@@ -23,15 +23,225 @@
...
+        See all commands by typing "?{hm|hangman} help".
```
Comment (le-potate) [inline cogs/games.py:84]: "See all categories"

File: cogs/games.py [hunk 43]
```diff
@@ -23,15 +23,225 @@
...
+
+    def help_str(self):
+        cat_list: str = ", ".join(
+            f"`{hm_cat}` (length: {len(self.hangman_dict[hm_cat][0])})"
+            for hm_cat in sorted(self.hangman_dict.keys()))
+        return ("rules: 5 wrong guesses are allowed, guesses "
+                "must be either the entire correct word or a "
+                "single letter (interpreted in a case insensitive manner)\n"
+                f"here is a list of valid category commands: {cat_list}\n"
+                "other commands are: `quit` (stops current game) "
```
Comment (le-potate) [inline cogs/games.py:69]: I'd write this like you did in the docsting (like `Quit an ongoing game by typing "?{hm|hangman} quit".`)

**Review by le-potate — CHANGES_REQUESTED** [review]

File: cogs/games.py (hunk 40 above)
Comment (namemcguffin) [reply cogs/games.py]: resolved in `d6bf98a`.

File: cogs/games.py (hunk 41 above)
Comment (namemcguffin) [reply cogs/games.py]: resolved in `d6bf98a`.

File: cogs/games.py (hunk 42 above)
Comment (namemcguffin) [reply cogs/games.py]: resolved in `d6bf98a`.

File: cogs/games.py (hunk 43 above)
Comment (namemcguffin) [reply cogs/games.py]: resolved in `d6bf98a`.

**Review by le-potate — APPROVED** [review]
